source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,723 @@
+        static assert(__traits(getAttributes, answerIs42)[0] == 42);
+    }
+
+    /**
+       Return mixin code for this `Function`.
+    */
+
+    string mixture() {
+        return join(
+            udas ~
+            attributeMixtureArray() ~
+            [
+                returnType,
+                name ~ "(" ~ parameterListMixtureArray.join(", ") ~ ")",
+            ], " ") ~
+            body_;
+    }
+
+    string[] parameterListMixtureArray() {
+        return map!(p => p.mixture)(parameters).array;
+    }
+
+    /**
+       Return the argument list as an array of strings.
+    */
+
+    const(string)[] argumentMixtureArray() {
+        return parameters.map!(p => p.name).array;
+    }
+
+    ///
+    unittest {
+        int add(int a, int b);
+        static assert(refract!(add, "add").argumentMixtureArray == [ "_0", "_1" ]);
+    }
+
+    /**
+       Return the argument list as a string.
+    */
+
+    string argumentMixture() {
+        return argumentMixtureArray.join(", ");
+    }
+
+    ///
+    unittest {
+        int add(int a, int b);
+        static assert(refract!(add, "add").argumentMixture == "_0, _1");
+    }
+
+    /**
+       Return the attribute list as an array of strings.
+    */
+
+    string[] attributeMixtureArray() {
+        with (FunctionAttribute) {
+            return []
+                ~ (static_ ? ["static"] : [])
+                ~ (attributes & pure_ ? ["pure"] : [])
+                ~ (attributes & nothrow_ ? ["nothrow"] : [])
+                ~ (attributes & property ? ["@property"] : [])
+                ~ (attributes & trusted ? ["@trusted"] : [])
+                ~ (attributes & safe ? ["@safe"] : [])
+                ~ (attributes & nogc ? ["@nogc"] : [])
+                ~ (attributes & system ? ["@system"] : [])
+                ~ (attributes & const_ ? ["const"] : [])
+                ~ (attributes & immutable_ ? ["immutable"] : [])
+                ~ (attributes & inout_ ? ["inout"] : [])
+                ~ (attributes & shared_ ? ["shared"] : [])
+                ~ (attributes & return_ ? ["return"] : [])
+                ~ (attributes & scope_ ? ["scope"] : [])
+                ~ (attributes & ref_ ? ["ref"] : [])
+                ;
+        }
+    }
+
+    ///
+    unittest {
+        nothrow pure int answer();
+        enum model = refract!(answer, "answer");
+        static assert(
+            model.attributeMixtureArray == ["pure", "nothrow", "@system"]);
+    }
+
+    /**
+       Return the attribute list as a string.
+    */
+
+    string attributeMixture() {
+        return attributeMixtureArray.join(" ");
+    }
+
+    ///
+    unittest {
+        nothrow pure int answer();
+        enum model = refract!(answer, "answer");
+        static assert(model.attributeMixture == "pure nothrow @system");
+    }
+}
+
+/**
+   A struct capturing all the properties of a function parameter.
+*/
+
+immutable struct Parameter {
```

Comment:
I think this struct seems like it should not be a builder like `Function`? As it is more of a data blob that is used to create or alter functions. Would a normal constructor do here? The fact that `refractParameter` maybe says something about providing refraction primitives in `Parameter`

Reply:
> I think this struct seems like it should not be a builder like `Function`? 

Oh yes it should! We want to be able to manipulate everything that composes the function. Openmethods does it [here](https://github.com/jll63/openmethods.d/blob/master/source/openmethods.d#L472-L475)